### PR TITLE
MVCCDeleteRange optimization.

### PR DESCRIPTION
**Pull request by petermattis**
I'm not terribly happy with copying code from MVCCIterate and
mvccPutInternal. I'll take a look later at whether that code duplication
can be eliminated while still maintaining the performance gains.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3146)

### CODE REVIEW

File: storage/engine/mvcc.go [hunk 1]
```diff
@@ -956,23 +956,125 @@ func MVCCMerge(engine Engine, ms *MVCCStats, key roachpb.Key, value roachpb.Valu
...
-		if err := MVCCDelete(engine, ms, kv.Key, timestamp, txn); err != nil {
-			return num, err
+	buf := putBufferPool.Get().(*putBuffer)
+	defer putBufferPool.Put(buf)
+	buf.value.Reset()
+	buf.value.Deleted = true
+
+	// We store encEndKey and encKey in the same buffer to avoid memory
+	// allocations.
+	encEndKey := mvccEncodeKey(buf.key[0:0], endKey)
```
Comment (tamird) [inline storage/engine/mvcc.go]: nit `[0:` is the same as `[:`

Comment (petermattis) [inline storage/engine/mvcc.go]: True, but does it matter? The generated code will be the same. I'll change the code this time, but I don't think we should be nit picking about such items.

Comment (tamird) [inline storage/engine/mvcc.go]: Sure, the only effect is grepability. I don't feel strongly; I made the comment only for the sake of consistency.

**Comment by bdarnell** [comment]
I'm very wary of the code duplication here. It looks easy for us to change `MVCC{Scan,Put,Delete}` but forget the corresponding code in `MVCCDeleteRange` (which is rarely used and therefore both less well tested and less likely to be worth this kind of optimization). I don't think we should optimize this unless we can do so without so much duplication.

**Comment by petermattis** [comment]
I don't like this level of optimization either. I'm tabling this PR while I experiment with moving the mvcc implementation from Go to C++.

**Comment by tamird** [comment]
Is this good to close, given recent discoveries regarding cgo performance?

**Comment by petermattis** [comment]
It's on my TODO list to revisit this change after some of the other MVCC changes settle.

**Pull request closed without merging**